packages/@aws-cdk/cdk/lib/resource-identifiers.ts

```diff
@@ -0,0 +1,50 @@
+import { ArnComponents } from './arn';
+import { CrossEnvironmentPhysicalArnToken, CrossEnvironmentPhysicalNameToken } from './cross-environment-token';
+import { Resource } from './resource';
+
+/**
+ * Construction properties for {@link ResourceIdentifiers}.
+ */
+export interface ResourceIdentifiersProps {
+  /**
+   * The {@link Resource} instance we're defining identifiers for.
+   */
+  readonly resource: Resource;
+
+  /**
+   * The ARN of the resource when referenced from the same CloudFormation template.
+   */
+  readonly resourceSimpleArn: string;
+
+  /**
+   * The name of the resource when referenced from the same CloudFormation template.
+   */
+  readonly resourceSimpleName: string;
+
+  /**
+   * The recipe for creating an ARN from a name for this resource.
+   */
+  readonly arnComponents: ArnComponents;
+}
+
+/**
+ * The identifiers (name and ARN) for a given L2.
+ * These should be only used inside the Construct Library implementation.
+ */
+export class ResourceIdentifiers {
+  public readonly arn: string;
+  public readonly name: string;
+
+  constructor(props: ResourceIdentifiersProps) {
+    this.arn = new CrossEnvironmentPhysicalArnToken(
```

Comment:
do we really need these two token classes? Can we make them internal to this function?

Reply:
They are already private to the `cdk` module, so I don't think making them internal to this function buys us much, and they're actually a little complex (at least for me), which suggests to me they should be separate.